dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ShowErrorWarningImplementerTest.java

```diff
@@ -71,173 +63,133 @@
-
-    @Autowired
-    private ShowErrorValidator errorImplementer;
-
-    @Autowired
-    private ShowWarningValidator warningImplementer;
-
-    private TrackerBundle trackerBundle;
-
-    @Override
-    protected void setUpTest()
-        throws IOException
-    {
-        renderService = _renderService;
-        userService = _userService;
+    private final static String ERROR_ON_COMPLETE_PREFIX = "Error on complete";
 
-        Map<Class<? extends IdentifiableObject>, List<IdentifiableObject>> metadata = renderService
-            .fromMetadata( new ClassPathResource( "tracker/event_metadata.json" ).getInputStream(), RenderFormat.JSON );
+    private final static String WARNING_ON_COMPLETE_PREFIX = "Warning on complete";
 
-        ObjectBundleParams params = new ObjectBundleParams();
-        params.setObjectBundleMode( ObjectBundleMode.COMMIT );
-        params.setImportStrategy( ImportStrategy.CREATE );
-        params.setObjects( metadata );
+    private final static String PROGRAM_STAGE_ID = "ProrgamStageId";
 
-        ObjectBundle bundle = objectBundleService.create( params );
-        ObjectBundleValidationReport validationReport = objectBundleValidationService.validate( bundle );
-        assertTrue( validationReport.getErrorReports().isEmpty() );
+    private final static String DATA_ELEMENT_ID = "DataElementId";
 
-        objectBundleService.commit( bundle );
+    private ShowWarningOnCompleteValidator warningOnCompleteImplementer = new ShowWarningOnCompleteValidator();
 
-        Program program = bundle.getPreheat().get( PreheatIdentifier.UID, Program.class, "BFcipDERJwr" );
+    private ShowErrorOnCompleteValidator errorOnCompleteImplementer = new ShowErrorOnCompleteValidator();
 
-        ProgramRule programRule = createProgramRule( 'B',
-            program );
-        programRule.setCondition( "true" );
-        programRuleService.addProgramRule( programRule );
-        ProgramRuleAction programRuleActionError = createProgramRuleAction( 'A', programRule );
-        programRuleActionError.setProgramRuleActionType( ProgramRuleActionType.SHOWERROR );
+    private ShowErrorValidator errorImplementer = new ShowErrorValidator();
 
-        programRuleActionError.setContent( CONTENT );
-        programRuleActionError.setData( DATA );
-        programRuleActionService.addProgramRuleAction( programRuleActionError );
+    private ShowWarningValidator warningImplementer = new ShowWarningValidator();
 
-        ProgramRuleAction programRuleActionErrorOnComplete = createProgramRuleAction( 'B', programRule );
-        programRuleActionErrorOnComplete.setProgramRuleActionType( ProgramRuleActionType.ERRORONCOMPLETE );
+    private TrackerBundle bundle;
 
-        programRuleActionErrorOnComplete.setContent( CONTENT );
-        programRuleActionErrorOnComplete.setData( DATA );
-        programRuleActionService.addProgramRuleAction( programRuleActionErrorOnComplete );
+    @Mock
+    private TrackerPreheat preheat;
```

Comment:
Mocking data is a bit controversial, in general, it should be avoided. I guess in this case, it's not worth to change.

Reply:
When we test and there is the Preheat involved is always difficult to go to the actual service to mock it.
In the end, we are using the Preheat as a service that give us the data from the DB in a performant way.

Reply:
Sorry Enrico, I placed the cursor on the wrong place. I meant this:

```
@Mock
private ProgramStage programStage;
```

Reply:
I agree, fixed